packages/hydrogen/src/cart/cart-api.ts

```diff
@@ -0,0 +1,161 @@
+export function createCartApi<TCustomMethods extends CustomMethodsBase>(
+  options: CartApiOptionsWithCustom<TCustomMethods>,
+): CartApiReturnCustom<TCustomMethods>;
+export function createCartApi<TCustomMethods extends CustomMethodsBase>(
+  options: CartApiOptions | CartApiOptionsWithCustom<TCustomMethods>,
+): CartApiReturn<TCustomMethods> {
+  const {requestHeaders, storefront, cartQueryFragment, cartMutateFragment} =
+    options;
+
+  // Default get cartId in cookie
+  const getCartId =
+    options.getCartId ||
+    (() => {
+      const cookies = parseCookie(requestHeaders.get('Cookie') || '');
+      return cookies.cart ? `gid://shopify/Cart/${cookies.cart}` : undefined;
+    });
+
+  // Default set cartId in cookie
+  const setCartId =
+    options.setCartId ||
+    ((cartId: string, headers: Headers) => {
+      headers.append('Set-Cookie', `cart=${cartId.split('/').pop()}`);
+    });
+
+  const mutateOptions = {
+    storefront,
+    getCartId,
+    cartMutateFragment,
+  };
+
+  const cartId = getCartId();
+  const cartCreate = cartCreateDefault(mutateOptions);
+
+  const methods: CartApiReturnBase = {
+    getFormInput,
+    get: cartGetDefault({
+      storefront,
+      getCartId,
+      cartQueryFragment,
+    }),
+    getCartId,
+    setCartId,
+    create: cartCreate,
+    addLine: async (cartInput: CartLinesAdd) => {
+      return cartId
+        ? await cartLinesAddDefault(mutateOptions)(cartInput)
+        : await cartCreate({
+            input: {lines: cartInput.lines},
+          });
+    },
+    updateLines: cartLinesUpdateDefault(mutateOptions),
+    removeLines: cartLinesRemoveDefault(mutateOptions),
+    updateDiscountCodes: async (cartInput: CartDiscountCodesUpdate) => {
+      return cartId
+        ? await cartDiscountCodesUpdateDefault(mutateOptions)(cartInput)
+        : await cartCreate({
+            input: {discountCodes: cartInput.discountCodes},
+          });
+    },
+    updateBuyerIdentity: async (cartInput: CartBuyerIdentityUpdate) => {
+      return cartId
+        ? await cartBuyerIdentityUpdateDefault(mutateOptions)({
+            buyerIdentity: cartInput.buyerIdentity,
+          })
+        : await cartCreate({
+            input: {buyerIdentity: cartInput.buyerIdentity},
+          });
+    },
+    updateNote: cartNoteUpdateDefault(mutateOptions),
+    updateSelectedDeliveryOption:
+      cartSelectedDeliveryOptionsUpdateDefault(mutateOptions),
+    metafieldsSet: async (cartInput: CartMetafieldsSet) => {
+      return cartId
+        ? await cartMetafieldsSetDefault(mutateOptions)(cartInput)
+        : await cartCreate({
+            input: {metafields: cartInput.metafields},
+          });
+    },
+    metafieldDelete: cartMetafieldDeleteDefault(mutateOptions),
+  };
+
+  if ('customMethods' in options) {
+    return {
+      ...methods,
+      ...(options.customMethods ?? {}),
+    };
+  } else {
+    return methods;
+  }
```

Comment:
Out of curiosity, wouldn't running always `return {...methods, ...options.customMethods}` make TypeScript happier? 🤔